crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -2517,6 +2523,118 @@ pub(crate) mod tests {
+        ).to_string());
+        assert_eq!(
+            converter
+                .parse_c8y_devicecontrol_topic(&original)
+                .await
+                .unwrap(),
+            vec![MqttMessage {
+                topic: Topic {
+                    name: "te/device/main///cmd/restart/c8y-mapper-16574089".into(),
+                },
+                payload: json!({"status":"init"}).to_string().into(),
+                qos: QoS::AtLeastOnce,
+                retain: true,
+            },]
+        );
+        assert_eq!(
+            converter
+                .parse_c8y_devicecontrol_topic(&redelivery)
+                .await
+                .unwrap(),
+            vec![]
+        );
+    }
+
+    #[tokio::test]
+    async fn custom_operations_are_deduplicated() {
+        let tmp_dir = TempTedgeDir::new();
+        let custom_op = r#"exec.topic = "my/custom/topic"
+        exec.on_fragment = "my_op"
+        exec.command = "/etc/tedge/operations/command ${.payload.my_op.text}""#;
+        let f = tmp_dir.dir("operations").dir("c8y").file("my_op");
+        f.with_raw_content(custom_op);
+        let (mut converter, _http_proxy) = create_c8y_converter(&tmp_dir).await;
+
+        let original = MqttMessage::new(&Topic::new_unchecked("my/custom/topic"), json!(
+            {"delivery":{"log":[],"time":"2025-03-05T08:49:24.986Z","status":"PENDING"},"agentId":"1916574062","creationTime":"2025-03-05T08:49:24.967Z","deviceId":"1916574062","id":"16574089","status":"PENDING","my_op":{},"description":"do something","externalSource":{"externalId":"test-device","type":"c8y_Serial"}}
+        ).to_string());
+        let redelivery = MqttMessage::new(&Topic::new_unchecked("my/custom/topic"), json!(
+            {"delivery":{"log":[{"time":"2025-03-05T08:49:24.986Z","status":"PENDING"},{"time":"2025-03-05T08:49:25.000Z","status":"SEND"},{"time":"2025-03-05T08:49:25.162Z","status":"DELIVERED"}],"time":"2025-03-05T08:49:25.707Z","status":"PENDING"},"agentId":"1916574062","creationTime":"2025-03-05T08:49:24.967Z","deviceId":"1916574062","id":"16574089","status":"PENDING","my_op":{},"description":"do something","externalSource":{"externalId":"test-device","type":"c8y_Serial"}}
+        ).to_string());
+        assert_ne!(
+            converter
+                .parse_json_custom_operation_topic(&original)
+                .await
+                .unwrap(),
+            vec![],
+            "Initial operation delivery produces outgoing message"
+        );
+        assert_eq!(
+            converter
+                .parse_json_custom_operation_topic(&redelivery)
+                .await
+                .unwrap(),
+            vec![],
+            "Operation redelivery is ignored by converter"
+        );
+    }
+
+    #[tokio::test]
+    async fn custom_operations_are_not_deduplicated_before_registration() {
```

Comment:
I struggle to understand what is checked by this test and how.

`CumulocityConverter.supported_operations` field is patched before the first request then restored before the second. Okay, but why?

Reply:
This tests is checking what happens if we receive a custom operation that is unrecognised and later get redelivered the custom operation after it is registered with the mapper. If the converter naïvely assumes that the operation is active after we first receive it, the de-duplication mechanism will ignore the redelivery. But since we haven't yet processed this message, we should process such a redelivery. This is obviously dependent on something sending a 500 message once the operation is registered, but that could be the case for some custom operation handling service.

The patching of the `supported_operations` was intended as an easy way of ensuring the registered operations are made clear, since I'm not trying to test how we update `supported_operations` in this case.

Reply:
This is clearer. Might be good to add this response as a comment to the test.

Reply:
Technically this is fine. But seeing the detailed discussion above, it felt like this case could have been better represented as an integration test in `tests.rs`, where we can better simulate the dynamic custom operation registration during the execution of the test.

Reply:
I think I want the opposite of that, @albinsuresh. The point of this being a unit test is I don't also want to test the operation registration logic at the same time. I've added a comment to explain what it is I'm trying to test.